common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/CollectionTree.scala

```diff
@@ -0,0 +1,94 @@
+package uk.ac.wellcome.models.work.internal
+
+import uk.ac.wellcome.models.work.internal.result._
+
+/**
+  * A CollectionTree contains a hierarchical archive or collection of works.
+  * Each work in the hierarchy contains a forward slash seperated path
+  * representing its position.
+  */
+case class CollectionTree(
+  path: String,
+  work: IdentifiedWork,
+  children: List[CollectionTree] = Nil
+) {
+
+  def size: Int =
+    children.map(_.size).sum + 1
+
+  def pathList: List[String] =
+    path :: children.flatMap(_.pathList)
+}
+
+object CollectionTree {
+
+  type Path = List[String]
+
+  /**
+    * Create a CollectionTree from a list of works. This errors given the following
+    * situations:
+    *
+    *   * If the works do not form a fully connected tree
+    *   * The given list contains works with duplicate paths.
+    *   * There are works which do not contain a Collection
+    *   * The list is empty
+    */
+  def apply(works: List[IdentifiedWork]): Result[CollectionTree] =
+    works
+      .map { work =>
+        work.data.collection match {
+          case Some(Collection(_, path)) => Right((tokenize(path), work))
+          case None =>
+            Left(
+              new Exception(
+                s"Work ${work.canonicalId} is not part of a collection"))
+        }
+      }
+      .toResult
+      .flatMap(apply(_, depth = 1))
+      .flatMap { tree =>
+        val pathList = tree.pathList
+        if (pathList.length < works.length) {
+          val unconnected =
+            works.map(_.data.collection.get.path).filterNot(pathList.toSet)
+          Left(new Exception(
+            s"Not all works in collection are connected to root '${tree.path}': ${unconnected
+              .mkString(",")}"))
+        } else if (pathList.length > pathList.toSet.size) {
+          val duplicates = pathList.diff(pathList.distinct).distinct
+          Left(
+            new Exception(
+              s"Tree contains duplicate paths: ${duplicates.mkString(",")}"))
+        } else
+          Right(tree)
+      }
+
+  private def apply(workMapping: List[(Path, IdentifiedWork)],
+                    depth: Int): Result[CollectionTree] =
+    workMapping.sortBy(_._1.length) match {
```

Comment:
You can unpack the tuple in a `sortBy`, makes it a bit more readable:

```suggestion
    workMapping.sortBy { case (path, _) => path.length } match {
```